File: packages/gg-sandbox/src/gg/docs.ts
/**
 * Find a module, function or type by searching, and close a documentation view.
 *
 * The system prompt names the modules and no function inside them; searching by keyword, or by
 * naming modules, lists what this session holds. A search answers with one-line briefs, and an
 * entry's full text is what a documentation view of its `key` holds.
 *
 * Every hit is a call this session can make: gg filters hits through the same permission the call
 * itself is checked against. The type checker reads the whole catalogue instead, and accepts a call
 * to a function this run withheld.
 */

import * as raw from "test-cabinet:gg/docs";
import { U32_MAX, arrayArg, call, opts, uint } from "../internal/errors.js";

/** Which of the three things a documentation entry is. */
export type DocKind =
  /** A module a program imports, whose functions live inside it. */
  | "module"
  /** A function a program calls. */
  | "function"
  /** A type a signature names. */
  | "type";

/** One entry a search matched: enough to choose from, and no more. */
export interface DocHit {
  /**
   * The fully-qualified name it is documented under, and the key a documentation view is opened by.
   */
  key: string;

  /** Whether this entry is a module, a function or a type. */
  kind: DocKind;

  /**
   * The module it lives in, written the way a program writes it.
   *
   * One module, always: the module that publishes a function, the module that declares a type,
   * and, for a module, itself. A type mentioned by several modules still reports the one place it is
   * defined. The value is accepted by the `modules` filter as written.
   */
  module: string;

  /** The name a program calls it by, the type's own name, or the module's path. */
  name: string;

  /** Its brief, and only its brief. Everything else written about it is what a view holds. */
  summary: string;
}

/** One page of search results, and the total behind it. */
export interface DocSearch {
  /** How many entries matched before paging, so a capped page is distinguishable from a whole one. */
  total: number;

  /** The offset this page starts at, echoed back, so paging needs nothing tracked on the side. */
  offset: number;

  /** The page itself, best first. */
  hits: DocHit[];
}

/**
 * Search the modules this session holds, every function it can call, and every type their signatures name.
 *
 * Best match first. A hit is a module a program imports, a function it calls, or a type a signature
 * names. Nothing gg offers is in scope until the program has imported the module the function lives
 * in.
 *
 * Matching is case-insensitive substring matching over names, signatures, briefs and detailed
 * descriptions. A hit whose own name matched ranks above one that mentions the word in a paragraph,
 * and the query is split on whitespace into separate terms.
 *
 * Every field is optional, and the filters compose with the query and with each other. Each filter
 * is an exact lookup rather than a ranking term, so `modules` with no query is those modules' whole
 * directory. A call carrying neither a query nor a filter is refused.
 *
 * The page comes back as a value and opens as a view labelled `search results`, which the next
 * search replaces.
 *
 * @ggop docs.search
 * @param options The words to look for, the filters, and the page. Every field is optional;
 * omitting the object entirely is refused.
 * @param options.query The words to look for, as one string. It may be left out when a filter says
 * what to look at instead.
 * @param options.modules The modules to look in, each named either the way a program writes it
 * (`gg.files`) or by gg's own id (`files`). Exact and case-insensitive, and several are a union: an
 * entry in any one of them is a hit. An empty list is no module filter at all.
 * @param options.type One type's own name, narrowing to that type and the functions whose signatures
 * mention it.
 * @param options.kind `"module"`, `"function"` or `"type"`, to see only one of them. Any other word
 * is refused.
 * @param options.offset How many hits to skip. Defaults to none.
 * @param options.limit How many hits to return: 20 by default, 100 at most, and zero is refused.
 * @returns one page of hits, best first, beside the `total` that says how many matched before
 * paging.
 * @throws `ApiError` with `invalid-argument` for a search carrying neither a query nor a filter, for
 * an unrecognised `kind`, and for a `limit` of zero.
 */
export function search(options?: {
  query?: string;
  modules?: string[];
  type?: string;
  kind?: DocKind;
  offset?: number;
  limit?: number;
}): DocSearch {
  const o = opts<{
    query?: string;
    modules?: string[];
    type?: string;
    kind?: DocKind;
    offset?: number;
    limit?: number;
  }>("search", options);
  const modules = arrayArg("search", "modules", o?.modules);
  const offset = uint("search", "offset", o?.offset, U32_MAX);
  const limit = uint("search", "limit", o?.limit, U32_MAX);
  return call(() => {
    const page = raw.search(o?.query, modules, o?.type, o?.kind, offset, limit);
    return {
      total: page.total,
      offset: page.offset,
      // The membrane declares `kind` a bare string, because the WIT record does; gg writes exactly
      // the three words this union carries, and the narrowing is here so a program branching on a
      // hit gets the closed set rather than `string`.
      hits: page.hits.map((hit) => ({ ...hit, kind: hit.kind as DocKind })),
    };
  });
}

/**
 * Close the documentation view opened under `key`.
 *
 * A key that is not open is not a failure. There is no cascade: closing a function's view leaves the
 * views of the types it named open, and closing a type's leaves the functions. A type closed here is
 * opened again by the next function that mentions it.
 *
 * @ggop docs.close
 * @param key The fully-qualified name the view was opened under, as a search hit reports it.
 * @returns how many views were closed, which is zero when that key was not open.
 * @throws `ApiError` with `unavailable` when this run did not enable the closing of documentation
 * views.
 */
export function close(key: string): number {
  // A `u32`, so already a `number` — nothing to widen on the way back.
  return call(() => raw.closeDocView(key));
}

/**
 * Close every documentation view at once.
 *
 * @ggop docs.close_all
 * @returns how many views went, which is zero when none was open.
 * @throws `ApiError` with `unavailable` when this run did not enable the closing of documentation
 * views.
 */
export function closeAll(): number {
  return call(() => raw.closeDocViews());
}

File: packages/gg-sandbox-csharp/src/Gg/Docs/Docs.cs
namespace Gg;

/// <summary>Search this surface's own documentation, and take a documentation view back out.</summary>
/// <remarks>
/// A search returns only the calls this run offers. The SDK compiles as one library, so a call the
/// search did not return still compiles and fails at run time with
/// <see cref="ApiErrorCode.Unavailable"/>, naming the capability the run withheld.
/// </remarks>
/// <ggmodule>docs</ggmodule>
public static partial class Docs
{
    /// <summary>Find the calls this run offers, by keyword or by module, and read the briefs.</summary>
    /// <remarks>
    /// <para>
    /// Matching is case-insensitive substring over names, signatures, briefs and detailed
    /// descriptions, so <c>"foobar"</c> finds <c>GetFoobar</c>. Hits are ranked by what matched: an
    /// entry whose own name carries the word comes before one that merely mentions it.
    /// </para>
    /// <para>
    /// The filters compose with the query and with each other, and <paramref name="modules"/> is a
    /// lookup rather than a search. A call with no query and no filter at all is refused. The page
    /// is returned as a value and is also opened as a view, and the next search replaces that view.
    /// </para>
    /// </remarks>
    /// <param name="query">
    /// The words to look for. Left out, the filters beside it are the whole of the search.
    /// </param>
    /// <param name="modules">
    /// The modules to look in, each by gg's own id (<c>"views"</c>) or by the path this arm writes
    /// it as (<c>"Gg.Views"</c>). Several are a union: an entry in any one of them is a hit. Exact
    /// and case-insensitive; a name no module has matches nothing rather than failing. Left out, or
    /// empty, no module filter at all.
    /// </param>
    /// <param name="type">
    /// One type's name, which narrows to that type and to the functions taking or returning it.
    /// </param>
    /// <param name="kind">
    /// Narrow to modules, to functions or to types. Left out, all three are searched.
    /// </param>
    /// <param name="offset">How many hits to skip. Left out, the page starts at the first.</param>
    /// <param name="limit">
    /// How many hits to return: 20 where it is left out, 100 at most, and zero refused.
    /// </param>
    /// <returns>one page of matches, best first, and the total behind it.</returns>
    /// <exception cref="ApiException">
    /// <see cref="ApiErrorCode.InvalidArgument"/> for a call with no query and no filter beside it,
    /// and for a <paramref name="limit"/> of zero.
    /// </exception>
    /// <ggop>docs.search</ggop>
    public static DocPage Search(
        string? query = null,
        string[]? modules = null,
        string? type = null,
        DocKind? kind = null,
        uint? offset = null,
        uint? limit = null)
    {
        Internal.Wire.Check(Internal.Native.SearchDocs(
            query,
            Internal.Wire.Or(modules),
            type,
            Internal.Wire.Word(kind),
            Internal.Wire.Slot(offset),
            Internal.Wire.Slot(limit),
            out var page,
            out var keys,
            out var kinds,
            out var hitModules,
            out var names,
            out var summaries));
        var hits = new DocHit[keys.Length];
        for (var index = 0; index < keys.Length; index++)
        {
            hits[index] = new DocHit(
                keys[index],
                Internal.Wire.Kind(kinds[index]),
                hitModules[index],
                names[index],
                summaries[index]);
        }
        return new DocPage(page[0], page[1], hits);
    }

    /// <summary>Take one documentation view back out of the context window, by its key.</summary>
    /// <remarks>
    /// A plain removal with no cascade: closing a function's view leaves the views of the types it
    /// mentions open, and closing a type's leaves the functions. A key that is not open closes zero
    /// rather than failing. A type closed here is opened again by the next function that mentions
    /// it.
    /// </remarks>
    /// <param name="key">The fully-qualified name the view was opened under.</param>
    /// <returns>how many views were closed, which is one unless it had already gone.</returns>
    /// <exception cref="ApiException">
    /// <see cref="ApiErrorCode.Unavailable"/> where this run does not offer the call.
    /// </exception>
    /// <ggop>docs.close</ggop>
    public static uint Close(string key)
    {
        Internal.Wire.Check(Internal.Native.CloseDocView(key, out var closed));
        return closed;
    }

    /// <summary>Take every documentation view out of the context window at once.</summary>
    /// <remarks>
    /// Only documentation views go, and having none open returns zero rather than failing.
    /// </remarks>
    /// <returns>how many views went.</returns>
    /// <exception cref="ApiException">
    /// <see cref="ApiErrorCode.Unavailable"/> where this run does not offer the call.
    /// </exception>
    /// <ggop>docs.close_all</ggop>
    public static uint CloseAll()
    {
        Internal.Wire.Check(Internal.Native.CloseDocViews(out var closed));
        return closed;
    }
}
